### PR TITLE
XYZ-35: Added Username to Compliance Export Queries

### PR DESCRIPTION
**Pull request by MusikPolice**
#### Summary
I added username to the queries that select compliance export data out of the database. This field is used in GlobalRelay exports. Also removed translations from some debug logging. I've talked to CH about this, and he's said that in the past, translating logs caused problems when attempting to help customers, and in this case the logs made it hard to execute the utils code from the enterprise repo, because it doesn't properly initialize the translations system.

#### Ticket Link
https://mattermost.atlassian.net/browse/XYZ-35

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json) updates

### CODE REVIEW

File: store/sqlstore/channel_member_history_store.go [hunk 1]
```diff
@@ -132,7 +133,8 @@ func (s SqlChannelMemberHistoryStore) getFromChannelMembersTable(startTime int64
 		SELECT DISTINCT
   			ch.ChannelId,
   			ch.UserId,
-  			u.email
+  			u.Email,
+            u.Username
```
Comment (jespino) [inline store/sqlstore/channel_member_history_store.go]: bad indentation

File: model/config.go [hunk 2]
```diff
@@ -1606,16 +1609,24 @@ func (s *PluginSettings) SetDefaults() {
 
 type MessageExportSettings struct {
 	EnableExport        *bool
+	ExportFormat        *string
```
Comment (MusikPolice) [inline model/config.go]: This used to just default to "actiance", but now that we have support for GlobalRelay, the user needs to be able to configure it for use.

File: utils/inbucket.go [hunk 3]
```diff
@@ -37,6 +38,12 @@ type JSONMessageInbucket struct {
 		Text string
 		HTML string `json:"Html"`
 	}
+	Attachments []struct {
+		Filename     string
+		ContentType  string `json:"content-type"`
+		DownloadLink string `json:"download-link"`
+		Bytes        []byte `json:"-"`
+	}
```
Comment (MusikPolice) [inline utils/inbucket.go]: Attachments are now supported for InBucket tests. This code will automatically download attachments and include them in the returned objects, so don't write a unit test that emails gigabyte sized attachments :)

File: utils/mail.go [hunk 4]
```diff
@@ -104,36 +104,56 @@ func TestConnection(config *model.Config) {
...
+	fromMail := mail.Address{Name: config.EmailSettings.FeedbackName, Address: config.EmailSettings.FeedbackEmail}
+	return sendMail(to, to, fromMail, subject, htmlBody, nil, nil, config)
+}
+
+// allows for sending an email with attachments and differing MIME/SMTP recipients
+func SendMailUsingConfigAdvanced(mimeTo, smtpTo string, from mail.Address, subject, htmlBody string, attachments []*model.FileInfo, mimeHeaders map[string]string, config *model.Config) *model.AppError {
+	return sendMail(mimeTo, smtpTo, from, subject, htmlBody, attachments, mimeHeaders, config)
+}
+
+func sendMail(mimeTo, smtpTo string, from mail.Address, subject, htmlBody string, attachments []*model.FileInfo, mimeHeaders map[string]string, config *model.Config) *model.AppError {
```
Comment (MusikPolice) [inline utils/mail.go]: GlobalRelay emails are quite a bit different than the normal emails that the system already sends. To support the various odds and ends that the system requires, I needed to overload the existing `SendEmailUsingConfig()` method to allow the user to specify a bunch of details.

Unfortunately, Golang doesn't support method overloading (I really miss Java some days), so this is the best I could do. I'm not 100% jazzed about the number of parameters in the advanced method signature, but it is called advanced, so...

File: utils/mail.go [hunk 5]
```diff
@@ -104,36 +104,56 @@ func TestConnection(config *model.Config) {
...
 		txtBody = ""
 	}
 
-	m := gomail.NewMessage(gomail.SetCharset("UTF-8"))
-	m.SetHeaders(map[string][]string{
-		"From":                      {fromMail.String()},
-		"To":                        {to},
+	headers := map[string][]string{
+		"From":                      {from.String()},
+		"To":                        {mimeTo},
```
Comment (MusikPolice) [inline utils/mail.go]: One of the major distinctions that GlobalRelay requires is differing the MIME `To` header and the SMTP `To` address. We always want to send the email to the GlobalRelay inbox, but we want to set the MIME `To` header to include all the of email addresses of the people that participated in the conversation.

File: utils/mail.go [hunk 6]
```diff
@@ -104,36 +104,56 @@ func TestConnection(config *model.Config) {
...
 		"Auto-Submitted":            {"auto-generated"},
 		"Precedence":                {"bulk"},
-	})
-	m.SetDateHeader("Date", time.Now())
+	}
+	if mimeHeaders != nil {
+		for k, v := range mimeHeaders {
+			headers[k] = []string{encodeRFC2047Word(v)}
+		}
+	}
```
Comment (MusikPolice) [inline utils/mail.go]: We also need the ability to add custom MIME headers

File: utils/mail.go [hunk 7]
```diff
@@ -104,36 +104,56 @@ func TestConnection(config *model.Config) {
...
+	m.SetHeaders(headers)
+	m.SetDateHeader("Date", time.Now())
 	m.SetBody("text/plain", txtBody)
 	m.AddAlternative("text/html", htmlMessage)
 
+	if attachments != nil {
+		for _, fileInfo := range attachments {
+			m.Attach(fileInfo.Path)
+		}
+	}
```
Comment (MusikPolice) [inline utils/mail.go]: And attachments too.

**Comment by MusikPolice** [comment]
The build failure here is a bit of a mystery to me. There's an [ongoing thread about the issue](https://pre-release.mattermost.com/core/pl/e365re5xybbamkxqnrxkaeubiy) in #Developers

**Comment by MusikPolice** [comment]
@grundleborg As far as I understand it from reading the GlobalRelay documentation, the emails get sent via a normal SMTP server to a special email account that GlobalRelay monitors. If it turns out that some customers prefer to use a custom SMTP server to talk to GlobalRelay (for instance, if they host Mattermost in one data centre and GlobalRelay in another, and don't want the emails crossing the open internet), then we may have to extract the SMTP config out of `EmailSettings` and add it to the existing `MessageExportSettings` object so that users can configure two separate SMTP servers to use.

**Comment by MusikPolice** [comment]
@jasonblais added export format to `app/diagnostics.go`. Don't think we should send the GlobalRelay email address, as that will be unique for every organization.

**Comment by MusikPolice** [comment]
@jasonblais No. `model/config.go` contains `(*MessageExportSettings) SetDefaults()`, which takes care of populating the fields in your config if you've left them unset. I don't really understand why we have to different ways of setting default values for config. I prefer the way that's in code, as it can be easily unit tested.

**Comment by jasonblais** [comment]
@MusikPolice I think there are admins who set the values via the config.json file. So if a setting is missing, they may not realize it and leave it unset.

I'm 0/5 though, as I don't fully understand it. As long as we're consistent with other features and config settings.

**Comment by MusikPolice** [comment]
**DO NOT MERGE** Until after the v4.7.0 release branch has been cut.

**Comment by MusikPolice** [comment]
The build just needs to be run again so that it pulls in the updates to the `enterprise` repo. The template snapshots have changed because I fixed a bug in that repo, and this build doesn't seem to have pulled in the changes.

**Comment by MusikPolice** [comment]
@jasonblais I queued your question about `default.json` for the next developer meeting. Here's [a thread about the topic](https://pre-release.mattermost.com/core/pl/usftm3rrmfbbfx3yffrktx74fc)

**Pull request merged**